pkg/execution/executor/executor.go

```diff
@@ -1856,8 +1867,15 @@ func (e *executor) HandleGeneratorResponse(ctx context.Context, i *runInstance,
 		)
 	}
 
-	groups := opGroups(resp.Generator).All()
-	for _, group := range groups {
+	groups := opGroups(resp.Generator)
+
+	if stepCount > 1 && i.md.ShouldCoalesceParallelism(resp) {
+		if err := e.smv2.SavePending(ctx, i.md.ID, groups.IDs()); err != nil {
+			return fmt.Errorf("error saving pending steps: %w", err)
+		}
+	}
```

Comment:
Here we save pending steps when handling the response but before handling each opcode.